### PR TITLE
[WIP] Dynamo traces through nn.modules

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87418

This causes dynamo to inline the code for nn.module.__call__ rather than
recording call_module nodes in the graph it will then hand to AOT to trace.

Note: AOT would trace through __call__ anyway, so this has no effect on the IR level
for downstream compilation/optimization.  Instead, it could cause modules that could
be traced wholesale into a dynamo graph to now cause graphbreaks, if they contain
unsupported behavior inside.

We would probably need to support an escape list of nn.modules that need to be captured
as nodes before we could land this, pending more investigation.

This change makes nn.module tracing match how user modules get traced, so in some ways
it's arguably a better policy.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jansel @lezcano @fdrocha

### CODE REVIEW

**Comment by Chillee** [comment]
I assume this is because FSDP or such monkey patches `nn.Linear`, and so you need to capture the operations within the module rather than simply the module as a whole?

I don't think this is a bad idea, but it does have some ramifications on whether we ever provide access to more of a `fx.symbolic_trace` like IR from Dynamo.

**Comment by wconstab** [comment]
> I assume this is because FSDP or such monkey patches nn.Linear, and so you need to capture the operations within the module rather than simply the module as a whole?

> I don't think this is a bad idea, but it does have some ramifications on whether we ever provide access to more of a fx.symbolic_trace like IR from Dynamo.


Yes, exactly my thoughts.  There are alternative ways to fix FSDP, but this may be a good thing anyway.  And this probably doesn't make FSDP bulletproof either, in the case we could not trace through a particular module.

**Comment by ezyang** [comment]
@wconstab any thoughts on also turning on this treatment for Python functions in torch.* namespace? It seems like this would be more uniform, and prevent weird situations where a function is treated differently if it gets refactored into PyTorch or not.

**Comment by ezyang** [comment]
For example, it would have prevented me from needing to write this diff

```
diff --git a/torch/nn/functional.py b/torch/nn/functional.py
index ebed3d9a60d..546c9119958 100644
--- a/torch/nn/functional.py
+++ b/torch/nn/functional.py
@@ -3910,7 +3910,11 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
             ]
         else:
             assert scale_factors is not None
-            output_size = [int(math.floor(float(input.size(i + 2)) * scale_factors[i])) for i in range(dim)]
+            from ..fx.experimental.symbolic_shapes import sym_int, sym_float
+            output_size = [
+                sym_int(math.floor(sym_float(input.size(i + 2)) * scale_factors[i]))
+                for i in range(dim)
+            ]
         scale_factors = None
 
     if antialias and not (mode in ("bilinear", "bicubic") and input.ndim == 4):
```

**Comment by wconstab** [comment]
I think it makes sense to trace through everything uniformly, down to aten if that's our convergence point.

Trouble is, I dunno if this is practically feasible (1), and if there is much of an issue with loss of 'module structure' information (2).   We can at least let CI and bench sweeps help figure out 1...

@SherlockNoMad @mlazos what's the fallout gonna be on our UX/debug tools if we have dynamo trace deeper into stuff and our dynamo/FX graphs are lower level?

**Review by jansel — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
> I think it makes sense to trace through everything uniformly, down to aten if that's our convergence point.
> 
> Trouble is, I dunno if this is practically feasible (1), and if there is much of an issue with loss of 'module structure' information (2). We can at least let CI and bench sweeps help figure out 1...
> 
> @SherlockNoMad @mlazos what's the fallout gonna be on our UX/debug tools if we have dynamo trace deeper into stuff and our dynamo/FX graphs are lower level?

For this model
```
class Block(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(3, 3, 3)
    def forward(self, x):
        x = self.conv(x)
        return x

class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.convs = torch.nn.Sequential(
            Block(),
        )
    def forward(self, x):
        x = self.convs(x)
        return x
```

Before
```
def forward(self, x : torch.Tensor):
        # Module stack: {'self_convs': <class 'torch.nn.modules.container.Sequential'>, 'self_convs_0': <class '__main__.Block'>, 'self_convs_0_conv': <class 'torch.nn.modules.conv.Conv2d'>}, File: /scratch/bahuang/work/repos/pytorch/temp/t.py:27, code: x = self.conv(x)
        self_convs_0_conv = self.self_convs_0_conv(x);  x = None
        return (self_convs_0_conv,)
```

After 
```
def forward(self, x : torch.Tensor):
        # Module stack: {'self_convs': <class 'torch.nn.modules.container.Sequential'>, 'self_convs_0': <class '__main__.Block'>}, File: /scratch/bahuang/work/repos/pytorch/temp/t.py:38, code: x = self.convs(x)
        self_convs_0_conv_weight = self.self_convs_0_conv_weight
        self_convs_0_conv_bias = self.self_convs_0_conv_bias
        self_convs_0_conv = self.self_convs_0_conv
        
        # Module stack: {'self_convs': <class 'torch.nn.modules.container.Sequential'>, 'self_convs_0': <class '__main__.Block'>, 'self_convs_0_conv': <class 'torch.nn.modules.conv.Conv2d'>}, File: /scratch/bahuang/work/repos/pytorch/torch/nn/modules/conv.py:463, code: return self._conv_forward(input, self.weight, self.bias)
        _conv_forward = self_convs_0_conv._conv_forward(x, self_convs_0_conv_weight, self_convs_0_conv_bias);  self_convs_0_conv = x = self_convs_0_conv_weight = self_convs_0_conv_bias = None
        return (_conv_forward,)        
```

If we further trace to aten
```
    def forward(self, arg0):
        # No stacktrace found for following nodes
        _param_constant0 = self._param_constant0
        _param_constant1 = self._param_constant1
        
        # Module stack: {'self_convs': <class 'torch.nn.modules.container.Sequential'>, 'self_convs_0': <class '__main__.Block'>, 'self_convs_0_conv': <class 'torch.nn.modules.conv.Conv2d'>}, File: /scratch/bahuang/work/repos/pytorch/torch/nn/modules/conv.py:463, code: return self._conv_forward(input, self.weight, self.bias)
        convolution_default = torch.ops.aten.convolution.default(arg0, _param_constant0, _param_constant1, [1, 1], [0, 0], [1, 1], False, [0, 0], 1);  arg0 = _param_constant0 = _param_constant1 = None
        return convolution_default
```
Module structure is correct, but call-site traceback is wrong.

**Comment by mlazos** [comment]
> I think it makes sense to trace through everything uniformly, down to aten if that's our convergence point.
> 
> Trouble is, I dunno if this is practically feasible (1), and if there is much of an issue with loss of 'module structure' information (2). We can at least let CI and bench sweeps help figure out 1...
> 
> @SherlockNoMad @mlazos what's the fallout gonna be on our UX/debug tools if we have dynamo trace deeper into stuff and our dynamo/FX graphs are lower level?

So I remember in the initial PT tooling brainstorm document that Xiaodong wrote, (sorry non-FB folk - https://docs.google.com/document/d/1xogK6gWHgJ3ZMIfUpMW5IO2F_JGZrAsCOFCKdBC8toI/edit#heading=h.12c9ogdw1cmm) that they wanted module nodes in the graph, and that this was a positive attribute over the older tracing system. I prefer not having to reconstruct this kind of information after it is lost from a design standpoint, but if there aren't other options this is okay.

Two questions: is the long term strategy for compiling distributed always going to be "trace it normally"? If you think we are going to eventually special case some of the distributed stuff in our stack anyway (ie insert a distributed module node in the graph that dynamo has configured), then I would say let's start special casing now.
This is influenced by my second question: does FSDP patch a lot of modules? If not, then maybe just trace through when needed?

**Comment by wconstab** [comment]
> Two questions: is the long term strategy for compiling distributed always going to be "trace it normally"? If you think we are going to eventually special case some of the distributed stuff in our stack anyway (ie insert a distributed module node in the graph that dynamo has configured), then I would say let's start special casing now.
This is influenced by my second question: does FSDP patch a lot of modules? If not, then maybe just trace through when needed?

> If there aren't other options though, I'm not strongly opposed, it might be good to consult some of the people who commented on the brainstorming document to make sure we can meet their needs through other means though.


Thinking longer term, we'll probably add support for tracing collective ops, be able to trace whole regions that we currently graph break on, and may update FSDP and/or use new distributed apis the compose better.  So I don't justify this change solely to unblock FSDP.  We can fix FSDP directly too, if we have to (i'm not entirely sure how hacky it would be).  But there are other proponents of this change for purposes outside of FSDP, which is what gave me the conviction to PR this.

> that they wanted module nodes in the graph, and that this was a positive attribute over the older tracing system

this goes directly to one of the questions i raised at the top.  I am not sure how much value we're adding by keeping it only to trace through it in AOT layer.  Is there a design somewhere that shows how we're gonna make use of the module structure from dynamo?  perhaps for export stack? @suo

**Comment by Chillee** [comment]
@wconstab @mlazos Reading between the lines on the doc, I don't think they actually need the module hierarchy in the graph they're dealing with, as opposed to debug information from the module hierarchy.

I mean... it's not like Dynamo preserves the module hierarchy in the FX graph anyways, except in a very limited sense.

**Comment by suo** [comment]
I am sort of late to this discussion, but dropping some thoughts about export since I was summoned 😛.

I see two major requirements for export that are of interest here:
- The export system should produce a well-defined IR that does not have references to Python (e.g. canonical ATen + whatever custom ops they had). This is important for hermetic serialization and backwards compatibility, which are both hard requirements for PyPER/Edge/others.
  - Our current strategy to achieve this is to go dynamo -> make_fx -> run passes to lower to canonical ATen. That seems largely unaffected by this PR, except inasmuch as this PR might change dynamo coverage/cause graph breaking.
- We should provide a "module-level" transformation API through FX where users can select/replace/move nodes belonging to a given module (or kind of module). This would let us replace the vast majority of what people are doing with FX symbolic tracer + leaf module customization today, and would probably be necessary for composition with  PyPER training optimizations.
  - Today we don't have anything for this. If we preserve enough information about where nodes came from (module hierarchy + tags that people place on modules), I think this wouldn't be too hard. It doesn't seem like this PR closes the door on that approach--but yes, we *do* want module structure preserved somehow.

**Comment by ezyang** [comment]
The PoR on module structure was to record module stacks on ops anyway...

**Comment by suo** [comment]
> The PoR on module structure was to record module stacks on ops anyway...

Yeah for sure, which I don't think this changes so sgtm. I think we'll probably want like some metadata propagation as well (e.g. you can store a tag on the module and it gets copied to the a `node["module_meta"]` dict or something) but haven't thought real hard about it.

**Review by mlazos — APPROVED** [review]
From what others have said I don't think it will be hard to recover the stack information for the modules that are traced. LGTM
